From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: gellert <gellert.hegyi@around.co>
Date: Tue, 19 May 2020 15:06:10 +0200
Subject: overwrites second channel data

This patch overwrites the second channel data using the first channel when sending audio out via webrtc.

Signed-off-by: gellert <gellert.hegyi@around.co>

diff --git a/pc/rtp_sender.cc b/pc/rtp_sender.cc
index aa268cef4594abdd2cb3a4a657d2d912a9c913c7..e0ecb80eefc8a360d908290da9f6e3a6572321b9 100644
--- a/pc/rtp_sender.cc
+++ b/pc/rtp_sender.cc
@@ -406,6 +406,20 @@ void LocalAudioSinkAdapter::OnData(
     size_t number_of_frames,
     absl::optional<int64_t> absolute_capture_timestamp_ms) {
   MutexLock lock(&lock_);
+
+  // @AROUND: Our modified WebRTC pipeline uses the second channel to encode metadata which
+  // needs to be stripped out before sending to the network. The below loop copies data from
+  // channel 0 to channel 1 instead of zeroing the channel to avoid a -6dB drop when subsequently
+  // downmixed.
+  if (number_of_channels == 2) {
+    auto data = (int16_t*)audio_data;
+    const size_t num_samples = number_of_channels * number_of_frames;
+    const size_t stride = 2;
+    for (size_t i = 0; i < num_samples; i += stride) {
+      data[i + 1] = data[i];
+    }
+  }
+
   if (sink_) {
     sink_->OnData(audio_data, bits_per_sample, sample_rate, number_of_channels,
                   number_of_frames, absolute_capture_timestamp_ms);
